Tools/build/check_warnings.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python3
+"""
+Parses compiler output with -fdiagnostics-format=json and checks that warnings exist
+only in files that are expected to have warnings.
+"""
+import sys
+from pathlib import Path
+import argparse
+import json
+import re
+
+def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
+    """
+    Extracts warnings from the compiler output when using -fdiagnostics-format=json
+
+    Compiler output as a whole is not a valid json document, but includes many json
+    objects and may include other output that is not json.
+    """
+
+    # Regex to find json arrays at the top level of the file in the compiler output
+    json_arrays = re.findall(r'\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*\])*\])*\]', compiler_output)
+    compiler_warnings = []
+    for array in json_arrays:
+        try:
+            json_data = json.loads(array)
+            json_objects_in_array = [entry for entry in json_data]
+            compiler_warnings.extend([entry for entry in json_objects_in_array if entry.get('kind') == 'warning'])
+        except json.JSONDecodeError:
+            continue  # Skip malformed JSON
```

Comment:
Is there a reason we want to skip malformed JSON?

Reply:
I found that sometimes the script will grab non-json objects in-between brackets so this just skips those if they can't be interpreted as json